### PR TITLE
[Issue #737] implement hidden timestamp column

### PR DESCRIPTION
**Pull request by gengdy1545**
Adding a hidden column into the pixels file and filtering out invisible tuples based on timestamp ordering.
This is compatible with the previous file format without such a hidden column.

### CODE REVIEW

**Comment by gengdy1545** [comment]
When loading data, a timestamp is assigned and grows normally (assuming it grows to k), but the timestamp corresponding to the query executed in trino-cli does not grow on the basis of k, so no data is read.

**Review by bianhq — CHANGES_REQUESTED** [review]

File: pixels-core/src/main/java/io/pixelsdb/pixels/core/vector/ColumnVector.java [hunk 1]
```diff
@@ -39,7 +39,7 @@
  * The fields are public by design since this is a performance-critical
  * structure that is used in the inner loop of query execution.
  */
-public abstract class ColumnVector implements AutoCloseable
+public abstract class ColumnVector implements AutoCloseable, Cloneable
```
[the remark this answers is not in the document]
Comment (gengdy1545) [reply pixels-core/src/main/java/io/pixelsdb/pixels/core/vector/ColumnVector.java:42]: Shallow copy for BinaryColumnVector type, read to columnVectors affects resultRowBatch.cols[i].writeIndex (PixelsRecordReaderImpl.java line1066). Haven't verified if it's a bug yet.

File: pixels-core/src/main/java/io/pixelsdb/pixels/core/reader/PixelsRecordReaderImpl.java [hunk 2]
```diff
@@ -137,6 +147,10 @@ public PixelsRecordReaderImpl(PhysicalReader physicalReader,
         this.footer = footer;
         this.option = option;
         this.transId = option.getTransId();
+        this.timestamp = option.getTimestamp();
+        this.includeHiddenColumn = this.timestamp != -1L;
```
Comment (bianhq) [inline pixels-core/src/main/java/io/pixelsdb/pixels/core/reader/PixelsRecordReaderImpl.java:151]: this.timestamp != -1L is not secure.

File: pixels-example/src/main/java/io/pixelsdb/pixels/example/core/TestColumnReader.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
+package io.pixelsdb.pixels.example.core;
```
Comment (bianhq) [inline pixels-example/src/main/java/io/pixelsdb/pixels/example/core/TestColumnReader.java:1]: Add license header.

File: proto/pixels.proto [hunk 4]
```diff
@@ -81,12 +83,18 @@ enum CompressionKind {
...
+    repeated ColumnStatistic columnStats = 3;
+    // file level hidden timestamp column statistics
+    optional ColumnStatistic hiddenColumnStats = 4;
     // row group information
-    repeated RowGroupInformation rowGroupInfos = 3;
+    repeated RowGroupInformation rowGroupInfos = 5;
     // row group statistics
-    repeated RowGroupStatistic rowGroupStats = 4;
+    repeated RowGroupStatistic rowGroupStats = 6;
+    // row group statistics for hidden timestamp columns
```
Comment (bianhq) [inline proto/pixels.proto:96]: It is better to add the fields for hidden columns at the end of the Footer. Thus, the new file format will be compatible with the old files.

**Review by bianhq — CHANGES_REQUESTED** [review]

**Review by bianhq — APPROVED** [review]
LGTM.

**Pull request merged**